pyro/poutine/handlers.py

```diff
@@ -115,82 +136,375 @@ def handler(fn=None, *args, **kwargs):
     return handler_decorator
 
 
-@_make_handler(BlockMessenger)
+@overload
 def block(
-    fn=None,
-    hide_fn=None,
-    expose_fn=None,
-    hide_all=True,
-    expose_all=False,
-    hide=None,
-    expose=None,
-    hide_types=None,
-    expose_types=None,
-): ...
+    fn: None = ...,
+    hide_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    expose_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    hide_all: bool = True,
+    expose_all: bool = False,
+    hide: Optional[List[str]] = None,
+    expose: Optional[List[str]] = None,
+    hide_types: Optional[List[str]] = None,
+    expose_types: Optional[List[str]] = None,
+) -> BlockMessenger: ...
+
+
+@overload
+def block(
+    fn: Callable[_P, _T] = ...,
+    hide_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    expose_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    hide_all: bool = True,
+    expose_all: bool = False,
+    hide: Optional[List[str]] = None,
+    expose: Optional[List[str]] = None,
+    hide_types: Optional[List[str]] = None,
+    expose_types: Optional[List[str]] = None,
+) -> Callable[_P, _T]: ...
+
+
+@_make_handler(BlockMessenger)
+def block(  # type: ignore[empty-body]
+    fn: Optional[Callable[_P, _T]] = None,
+    hide_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    expose_fn: Optional[Callable[["Message"], Optional[bool]]] = None,
+    hide_all: bool = True,
+    expose_all: bool = False,
+    hide: Optional[List[str]] = None,
+    expose: Optional[List[str]] = None,
+    hide_types: Optional[List[str]] = None,
+    expose_types: Optional[List[str]] = None,
+) -> Union[BlockMessenger, Callable[_P, _T]]: ...
+
+
+@overload
+def broadcast(
+    fn: None = ...,
+) -> BroadcastMessenger: ...
+
+
+@overload
+def broadcast(
+    fn: Callable[_P, _T] = ...,
+) -> Callable[_P, _T]: ...
 
 
 @_make_handler(BroadcastMessenger)
-def broadcast(fn=None): ...
+def broadcast(  # type: ignore[empty-body]
+    fn: Optional[Callable[_P, _T]] = None,
+) -> Union[BroadcastMessenger, Callable[_P, _T]]: ...
+
+
+@overload
+def collapse(
+    fn: None = ...,
+    *args: Any,
+    **kwargs: Any,
+) -> CollapseMessenger: ...
+
+
+@overload
+def collapse(
+    fn: Callable[_P, _T] = ...,
+    *args: Any,
+    **kwargs: Any,
+) -> Callable[_P, _T]: ...
 
 
 @_make_handler(CollapseMessenger)
-def collapse(fn=None, *args, **kwargs): ...
+def collapse(  # type: ignore[empty-body]
+    fn: Optional[Callable[_P, _T]] = None,
+    *args: Any,
+    **kwargs: Any,
+) -> Union[CollapseMessenger, Callable[_P, _T]]: ...
+
+
+@overload
+def condition(
+    data: Union[Dict[str, "torch.Tensor"], "Trace"],
+) -> ConditionMessenger: ...
```

Comment:
Some handlers have arguments that are not optional so `fn` has to be required in the signature. This is the trick with overloading to have `fn` as optional. Based on the signature mypy can figure out which type annotations to use.